♻️ Refactoring

ItsMii176 opened the pull request.

ItsMii176 committed:
➕ add TypeScript seeding script for Supabase

ItsMii176 committed:
🗃️ Implement complete database seeding system with Drizzle ORM

- Add PostgreSQL schema with foreign key relationships (organizations, students, internships, complete)
- Implement reproducible data seeding with Faker.js
- Create database cleanup utilities for clean development workflow
- Add comprehensive data display functions with relational queries
- Configure Drizzle Kit for database migrations and schema management
- Set up TypeScript compilation and npm scripts for seed/clean/query operations
- Add proper .gitignore rules for node_modules and build artifacts
- Ensure 1:1 student-internship relationships for independent data modeling

ItsMii176 committed:
✨ Refactor internships to use database and API integration

Replaces usage of static fake data with dynamic data fetched from a new /api/internships endpoint backed by a PostgreSQL database via Drizzle ORM. Adds database schema, query utilities, data converters, and React hooks for fetching and displaying internships. Updates UI components to use new types and data flow, adds skeleton loaders, and improves error handling and sorting. Cleans up seed scripts and removes unnecessary console logs.

ItsMii176 committed:
🔧 Parser-DB integration with enhanced error handling and field normalization

- Connected the parser to the database, enabling real-time data retrieval
- Implemented HTTP 500 error response when a database read operation fails
- Added a clear error message when no internship/stage entries are found
- Improved normalization logic across various fields to ensure consistent formatting and cleaner data input/output

ItsMii176 committed:
🧹Remove organization type update scripts

ItsMii176 committed:
:lipstick: Remove extra background and spacing in NotFound page

ItsMii176 committed:
♻️  Refactor DB queries to use Supabase client

Replaces Drizzle ORM with Supabase client for fetching internship, student, and organization data. Updates query logic, sorting, and data transformation to match Supabase's API and response structure. Improves error handling and ensures environment variables are checked for Supabase configuration.

ItsMii176 committed:
➕  Implement infinite scroll for internships list

🚧 Sorting via Supabase API still needs to be implemented

ItsMii176 committed:
✨Display internship ID in InternshipCard

ItsMii176 committed:
✨ Add internship details dialog and postal code support

romainbourdain committed:
🔥 Remove excel file

romainbourdain committed:
📝 Admin pages comments

ItsMii176 committed:
✨ Enhance select and internship list with sorting and loading states

Refactored the Select component to support reselecting the current value and improved className consistency. Updated internship list and header to handle loading states and display skeletons. Enhanced sorting functionality in SortInternshipButton to allow toggling sort order by reselecting the same option, and improved UI feedback for sorting direction.

ItsMii176 committed:
Merge branch 'intership-page-features' of https://github.com/info-telecom-strasbourg/chronos-nova into interships-page-features

romainbourdain committed:
Merge branch 'intership-page-features' of github.com:info-telecom-strasbourg/chronos-nova into intership-page-features

romainbourdain committed:
🌱 Refactor backend

romainbourdain committed:
🗃️ Implement drizzle schema and seeding

romainbourdain committed:
🐛 Implement drizzle schema and seeding

romainbourdain committed:
🗃️ Add academic year to internship table

ItsMii176 committed:
✨ Add search functionality to internship list

Introduced a debounced search input for internships using a new InternshipSearch component. Updated InternshipHeader and InternshipListClient to support searching, and added the use-debounce dependency.

romainbourdain committed:
💩 Solved errors

romainbourdain committed:
🔀 Merge with latest branch

romainbourdain committed:
♻️ Refactoring home page

romainbourdain committed:
📱 Fix spacing issues

ItsMii176 committed:
✨ Add custom error and not-found pages

ItsMii176 committed:
🐛 Fixed graphical bugs and implemented missing features following refactoring.

ItsMii176 committed:
✨ Add admin mode with route toggle and UI updates

Introduces an admin mode toggle in the navbar, a useAdminMode hook, and updates internship card and skeleton components to show admin actions (edit, delete) when in admin mode. Adds InternshipList and integrates it into the admin page. InternshipHeader now displays an 'Add Internship' button for admins.

romainbourdain committed:
🔀 Merge with remote branch

romainbourdain committed:
⚡️ Fix infinite scroll

ItsMii176 committed:
✨ Add internship creation form and related UI components

Implemented a full-featured internship creation form with validation using React Hook Form and Zod. Added UI components for autocomplete, textarea, and command palette. Introduced form sections for organization, internship details, and student information. Updated the admin interface to link to the creation page.

ItsMii176 committed:
♻️  Refactor admin buttons to use asChild and improve schema

ItsMii176 committed:
✨ Add debounced search input to internship search

ItsMii176 committed:
✨ Add skeleton loading state for internships page

ItsMii176 committed:
✨ Add pagination skeletons to internship infinite scroll

Introduces the InternshipPaginationSkeleton component to display loading skeletons when fetching additional internship pages. Replaces the spinner with skeleton cards for a more consistent loading experience during infinite scroll.

ItsMii176 committed:
♻️ Refactor student major/option selection with dynamic options

Introduces an academic structure to dynamically generate available majors and options based on the selected academic year. Updates the internship schema to validate major/option combinations and require options only when applicable. Refactors the student section UI to use the new dynamic logic, disables fields until prerequisites are selected, and adds contextual hints for improved user guidance.

ItsMii176 committed:
✨ Add internship edit form and refactor form sections

Introduces an edit form for internships with pre-filled data and server-side fetching. Refactors form sections to use React Hook Form's Controller for better state management, updates organization type values, and enables navigation to the edit page from the internship card. Also adds a server action to fetch internship details by ID.

ItsMii176 committed:
✨ Add toast on disabled autocomplete click (major, option)

romainbourdain commented on app/admin/create/page.tsx:
Evite de faire des pages qui ont un rendu côté client. Fais plutôt une page côté serveur asynchrone et un composant CreateInternshipForm qui contient le formulaire et qui lui est côté client.

Petite remarque : si tu te débrouilles bien, le formulaire de création et d'édition peuvent être dans le même composant. Normalement ils sont assez similaires et tu as juste à changer quelques valeurs et les default values.

romainbourdain commented on app/layout.tsx:
```diff
@@ -13,12 +14,13 @@ export const metadata: Metadata = {
+          <main className="mx-auto flex w-full max-w-screen-md flex-1 flex-col px-5 py-3">
             {children}
           </main>
+          <Toaster position="top-center" theme="system" />
```
Mets le Toaster dans le composant Providers directement. Il est fait pour regrouper tout ce genre de Composant qui setup des librairies.

romainbourdain commented on app/layout.tsx:
```diff
@@ -2,6 +2,7 @@ import { Providers } from "@/components/providers";
 import "@/styles/globals.css";
 import type { Metadata } from "next";
 import type { LayoutParams } from "@/types/next";
+import { Toaster } from "sonner";
```
Importe toaster depuis le composant de shadcn pour avoir les couleurs de shadcn

romainbourdain commented on src/components/layout/navbar.tsx:
```diff
@@ -15,6 +16,7 @@ export const Navbar = () => {
           <span className="font-semibold text-lg">Chronos</span>
         </Link>
         <div className="flex items-center gap-2">
+          <RouteToggle />
```
Le nom de ce composant n'est pas très clair, peut être plutôt l'appeler AdminToggle

romainbourdain commented on src/features/form/edit-form-client.tsx:
Même remarque que pour le form de création : essaie de fusionner les deux car une grande partie de la logique est la même. Tu peux mettre une prop edit qui est un booléen qui permet de choisir quel type de formulaire tu veux, et mettre defaultValues en argument pour préremplir le formulaire edit

romainbourdain commented on src/features/form/internship-details-section.tsx:
```diff
@@ -0,0 +1,102 @@
+            className={errors.subject ? "border-destructive focus-visible:ring-destructive" : ""}
+            placeholder="Description détaillée du sujet de stage"
+          />
+          {errors.subject && <p className="text-destructive text-sm">{errors.subject.message}</p>}
```
Regarde dans la doc de shadcn mais normalement tu dois utiliser les composant du fichier Form :
- FormField et FormItem pour englober tout ton champ
- FormLabel pour le label
- FormControl pour encapsuler uniquement ton TextArea ou Input ou etc
- FormMessage pour ajouter un message en cas d'erreur

Lien de la doc : https://ui.shadcn.com/docs/components/form

romainbourdain commented on src/features/form/options.ts:
Eventuellement on pourra mettre dans le backend ce genre de données pour que si une filière est créée ou disparait, on n'est pas à modifier le code source du site et l'on puisse juste ajouter des données à la db

romainbourdain commented on src/features/form/organization-section.tsx:
Même remarque que pour InternshipDetailsSection, il faut que tu utilises les composants de shadcn pour les form

romainbourdain commented on src/features/form/student-section.tsx:
Même remarque que pour InternshipDetailsSection, il faut que tu utilises les composants de shadcn pour les form

romainbourdain commented on src/hooks/use-infinite-scroll.ts:
Je crois que c'est moins qui aie laissé ce fichier mais je ne crois pas qu'on l'utilise encore. Tu peux le supprimer

romainbourdain commented on src/lib/actions/internship.ts:
Deux points sur ce fichier :
- Un fichier accent doit commencer par "use server"
- Une action est faite lorsque tu apportes une mutation sur les données (create, edit, delete). Ce n'est pas le cas des get et il faut éviter de créer des server actions pour ces requête
- En fait il y a aussi un troisième point : le fichier ne devrait pas se trouver ici mais plutôt dans features. Tu peux le nommer internship.action.ts

romainbourdain commented on src/lib/scripts/date.ts:
J'ai laissé ton ancienne version en commentaire mais si tout fonctionne comme avant tu peux retirer les commentaires

ItsMii176 committed:
♻️ Refactor Toaster and rename RouteToggle to AdminToggle

ItsMii176 committed:
♻️ Refactor internship form to use unified component

Replaces separate create and edit internship forms with a single reusable InternshipForm component. Updates related pages and form sections to use React Hook Form's FormField API for improved consistency and maintainability. Removes unused hooks and the old edit form client component.

ItsMii176 committed:
🚚 Move internship actions to features directory

ItsMii176 committed:
✨ Add state enum and field to internships schema

ItsMii176 committed:
✨ Add admin tabs and state management for internships

Introduces admin tab navigation with counts for active, pending, and deleted internships. Adds new admin pages for deleted and pending internships, updates infinite scroll and queries to support filtering by state, and enhances the internship card with state badges and context-aware admin actions. Also adds UI components for tabs and badges, and updates dependencies to include @radix-ui/react-tabs.

ItsMii176 committed:
💄 Add vertical spacing to admin page containers

ItsMii176 committed:
♻️ Refactor admin action buttons layout in InternshipCard

ItsMii176 committed:
✨ Add admin actions for internships and improve form handling

Introduces admin actions (approve, soft delete, restore) for internships with confirmation dialog. Refactors internship form to handle create and update operations with feedback and navigation. Updates internship card to use new admin actions and dialog. Adds corresponding queries for admin actions in the backend.

ItsMii176 committed:
💄 Add error prop to select fields in form sections

ItsMii176 committed:
♻️ Refactor admin dialogs and actions, add hard delete

Replaces custom admin dialog and actions with a reusable ConfirmDialog component based on a new AlertDialog abstraction. Removes fake data and related utilities, and updates internship actions to support hard delete with confirmation. All internship mutations now trigger admin page revalidation. Cleans up unused files and improves admin card action UX.

ItsMii176 committed:
:bug: Fix pluralization logic in internship count label

ItsMii176 committed:
✨ Add Excel import feature with exceljs dependency

Introduced a new Excel import dialog component and related types to support importing data from Excel files. Added the 'exceljs' package to dependencies for Excel file handling. Updated the internship form to integrate the new import functionality.

ItsMii176 committed:
♻️ Refactor Excel import and parser for 2A internships

Replaces legacy parser and Supabase inserter with a new server action for Excel import, supporting only '2A - Récap. stage' and 'Stage substitution' sheets. Adds UI validation for supported sheets, integrates a new Checkbox component, and updates data normalization and parsing logic for improved robustness and type safety. Removes obsolete parser CLI and inserter files, and introduces transformation helpers for 2A data. Updates form logic to handle Excel import and display user feedback.

ItsMii176 committed:
♻️  Refactor parser utilities and remove unused modules

Consolidated all parsing, normalization, and transformation functions into a new 'parser-utils.ts' file. Removed redundant and fragmented modules: data-normalizer.ts, excel-validator.ts, functions.ts, and parser-2A-transformations.ts. Updated imports in parser-internship-2A.ts and parser-substitution-internship.ts to use the new utility module, improving maintainability and code organization.

ItsMii176 committed:
🐛  Show only visible internships in infinite scroll

Updated the internships count and infinite scroll to use only 'visible' internships by passing the 'visible' state to getInternshipsCount and InternshipInfiniteScroll. This ensures that only internships marked as visible are displayed to users.

ItsMii176 committed:
💄Refactor admin tabs to use buttons with links

Replaces the Tabs component with Button components that use Next.js Link for navigation. Removes useRouter and related tab change handler, simplifying the tab navigation logic and improving accessibility.

ItsMii176 committed:
♻️ Refactor major and option display with alias and name

Updated internship details dialog to display major and option using both alias and full name, removing previous mapping functions. Added a new query to fetch internships with joined major and option details. Refactored correspondence utilities to provide formatting helpers and updated major/option string mappings.

ItsMii176 committed:
♻️ Standardize major and option aliases across app

Refactored major and option alias handling to use consistent transformation functions throughout the codebase. Updated parsing utilities, database upserts, and display logic to ensure aliases from Excel and user input are mapped to canonical forms. Added formatting helpers for city and country names, and extended mapping tables for legacy alias support.

ItsMii176 committed:
💄Improve display of major and option in internship views

Updated internship card and details dialog to better handle and display student major and option, including mapping aliases to full names and hiding 'AUCUNE' options.

ItsMii176 committed:
♻️ Refactor internship parser and refresh after form submit

Refactored the internship Excel parser to use a generic row parsing function, reducing code duplication and improving maintainability. Also added a router.refresh() call after successful internship form submission to ensure the UI updates accordingly.

ItsMii176 committed:
💄  Add organization type labels and date formatting to internship card

ItsMii176 committed:
♻️  Refactor major/option normalization and enrich display labels

Replaces static mapping files for majors and options with dynamic normalization and enriched label functions in parser-utils. Centralizes normalization logic in a new stage-normalizer utility, updates all usages to rely on normalized short values, and removes legacy mapping files. Also adds more country options and improves organization type and country normalization for both import and form flows.

ItsMii176 committed:
🐛 Fix insertion of name fields for major and option when saving an internship

ItsMii176 committed:
💄 Add 'pm' option for Physique et Modélisation

ItsMii176 committed:
✨ Add search functionality to internship infinite scroll

Refactored the infinite scroll and query logic to support searching internships by organization name or subject. The search is triggered only for queries with at least 2 characters. Updated the UI to debounce and filter search input, and removed the need to pass totalItems as a prop, instead deriving it from query results.

ItsMii176 committed:
✨ Add Master options to academic structure and parser utils

ItsMii176 committed:
♻️ Centralize mappings and normalization logic

Refactor to move all mappings and normalization functions for majors, options, countries, and organization types into a new parser/mappings.ts module. Update all usages across form options, infinite scroll components, parser utilities, and normalization utilities to use the new centralized functions and mappings. This improves maintainability, consistency, and extensibility of label and parsing logic throughout the codebase.

ItsMii176 committed:
🧹Remove unused imports and comments from parser-utils

ItsMii176 committed:
✨ Show message when no internships are found

ItsMii176 committed:
🐛 Fix bug from organization type mappings for parsing and display

Separated organization type mappings into display and parsing objects. Updated parseOrganizationType to use the new alias mapping for parsing, improving clarity and maintainability.

ItsMii176 committed:
💄 Refactor admin internship skeletons and remove total count

Removed the unused totalInternships count from admin pages and updated the skeleton components to accept a configurable number of admin action buttons.

ItsMii176 committed:
⚰️ Remove student name and internship confidential fields

Eliminates firstName and lastName fields from the student schema and related code, and removes the confidential field from the internship schema and all associated logic. Updates forms, parsers, normalization utilities, and database interactions to reflect these changes, simplifying data handling and reducing unnecessary fields.

ItsMii176 committed:
⚰️Remove support for 'Stage substitution' sheet in Excel import

Excel import now only supports the '2A - Récap. stage' sheet. All references to 'Stage substitution' have been removed from the dialog, action, and parser logic. The parser for 'Stage substitution' has also been deleted.

ItsMii176 committed:
⚰️Remove actions prop from internship skeleton components

ItsMii176 committed:
♻️ Refactor admin internship create/edit to dialogs

Removed admin internship create and edit pages, replacing them with dialog components for creating and editing internships. Updated admin tabs and internship card to use the new dialog-based workflow. Improved form validation and refactored schema to handle undefined values. Added new header component for admin internship list.

ItsMii176 committed:
⬆️ Update dependencies and scripts in package.json

Upgraded multiple dependencies and devDependencies to their latest versions, including Next.js, React, Zod, Drizzle ORM, and others. Removed Turbopack from the dev script.

ItsMii176 committed:
♻️ Replace AutoComplete with Select and Combobox components

Removed the AutoComplete component and replaced its usage with Select and Combobox components for improved consistency and UI. Added new Popover and Combobox components, updated Excel import dialog to require academic year and start row for specific sheets, and adjusted related types and actions to support academic year selection. Updated styles and error handling for form fields and Excel import workflow.

ItsMii176 committed:
♻️  Refactor internship form option handling and validation

Standardizes the handling of student options in internship forms by always providing an 'aucune' option when none are available, updates schema validation to require a non-empty option, and improves error feedback for required fields. Removes unused FormMessage components and simplifies form field logic. Updates Excel import and parser to pass academic year and startRow consistently.

ItsMii176 committed:
🚨 Fix grid class order in student section

ItsMii176 committed:
♻️ Refactor data model to support null and unknown values

This commit updates database schemas, parser utilities, and UI components to consistently handle null or unknown values for internship, student, and organization fields. It introduces '__inconnu__' as a placeholder for unknown majors and options, makes relevant fields nullable in both the database and TypeScript types, and adds display helpers to show '??' for missing data. The commit also removes the obsolete date formatting utility and ensures all data normalization and validation logic is compatible with the new approach.

ItsMii176 committed:
🔨 Update importantCells indices in hasSignificantContent

Changed the importantCells array in hasSignificantContent to check columns 5 and 16 (Titre stage and Dates stages) instead of the previous indices. Also removed the check for '??' as a non-significant value.

ItsMii176 committed:
♻️ Refactor internship validation and improve import handling

Moved internship validation schemas to a new stage-validation module and updated all usages. Improved Excel import to count and report badly imported internships (with missing or unknown data). Updated UI and logic to consistently treat '__inconnu__' as unknown for majors/options, and improved display and validation accordingly. Removed the old internship-validation.schema.ts file.

ItsMii176 committed:
🚨Fix CSS linting warnings

ItsMii176 committed:
♻️ Refactor country handling to use i18n-iso-countries

Replaces hardcoded country mappings with dynamic generation using the i18n-iso-countries package and French locale. Adds utility functions for country normalization, lookup, and validation in src/lib/utils/countries.ts. Updates form options and parser logic to use the new country utilities, improving maintainability and internationalization.

ItsMii176 committed:
♻️ Refactor mappings and parsing utilities to lib directory

Centralizes all mappings, parsing, and formatting utilities into the /lib directory for better modularity and maintainability. Moves mapping constants and types to src/lib/const/mappings.ts, parsing functions to src/lib/utils/parsers.ts, formatting functions to src/lib/utils/formatters.ts, and string normalization to src/lib/utils/string-normalizer.ts. Updates all relevant imports and re-exports in feature and form files to use the new structure.

ItsMii176 committed:
🚸 Adjust infinite scroll UI and inView root margin

Set a 200px root margin for useInView to trigger loading earlier.

ItsMii176 committed:
💄 Add z-50 to navbar for higher stacking context

Added the z-50 class to the navbar header to ensure it appears above other elements when scrolling. This helps prevent the navbar from being overlapped by other components with lower z-index values.

ItsMii176 committed:
🚸 Add scroll handling to CommandList in Combobox

ItsMii176 committed:
🔨 Add ESE option to academic structure and mappings

ItsMii176 committed:
🔥 Remove sheet name restriction in Excel import

The Excel import dialog and action now allow importing from any selected sheet, not just '2A - Récap. stage'. Validation and UI logic have been updated to require academic year and start row for all sheets, and the parser is now called for all selected sheets without filtering by name.

ItsMii176 committed:
🚸 Improve import message for zero imported items

Adds a specific message when no stages are imported, prompting users to check their data. This enhances user feedback during the Excel import process.

ItsMii176 committed:
✨ Add internship hash for duplicate detection and validation

Introduces an internshipHash field to the internship schema and updates all related logic to generate and store a hash for each internship based on key fields. Also adds object-hash and its types as dependencies.

ItsMii176 committed:
💄Refactor display logic for major and option fields

Simplifies and unifies the logic for displaying the student's major and option in both internship-card and internship-details-dialog components. Now, the option is only shown if it exists and is not 'aucune', and unknown values are consistently represented as '??'.

ItsMii176 committed:
✨ Add invalid and duplicate internship detection

Introduces an 'isInvalid' boolean field to the internship schema and updates creation, import, and update logic to set this flag based on validation. Adds server-side duplicate detection utilities and enriches draft internships with an 'isDuplicate' flag. Updates UI to display badges for invalid or duplicate internships in the admin view.

ItsMii176 committed:
♻️ Refactor internship badges and duplicate detection logic

Replaces StageBadges with a new InternshipBadges component that unifies state and validation badges for internships. Updates usage in internship-card to use the new component. Renames duplicate detection utility functions for clarity and updates their usage in internship.query and duplicate-detection-server.

ItsMii176 committed:
🐛 Fix badge logic for draft internships in search results

ItsMii176 committed:
✨ Add Excel import summary dialog and row count tracking

Introduces an ExcelImportSummaryDialog component to display a summary after importing internships from Excel. Updates the import logic and parser to track the number of rows read, imported, badly imported, and duplicates. Adjusts types and dialog flow to show the summary after a successful import.

ItsMii176 committed:
🚸 Show loading dialog during internship import

Added a modal dialog with a spinner and message to indicate progress while importing internships from Excel. The dialog uses @radix-ui/react-visually-hidden for accessibility and prevents closing while the import is in progress. Updated dependencies to include @radix-ui/react-visually-hidden.

ItsMii176 committed:
🐛 Fix toast message when approval is false

ItsMii176 committed:
✨ Add bulk delete actions for admin internships

Introduces DeleteAllButton and InternshipConfirmDialogs components to allow admins to bulk delete draft or deleted internships with confirmation dialogs. Updates admin pages and headers to use these new controls. Refactors internship-card to use the shared confirmation dialog logic. Adds deleteAllDrafts and deleteAllDeleted actions to handle bulk operations in the data layer.

ItsMii176 committed:
🚸 Add academic data validation and fix in stage normalization

Introduces a utility to validate and correct academic year, major, and option coherence. Updates stage normalization to use this validation when importing from Excel, ensuring data consistency. Also refines badly imported stage detection to include invalid student options.

ItsMii176 committed:
🐛 Change internship restore to set state as draft

Restoring an internship now sets its state to 'draft' instead of 'visible', and updates the success message to indicate the internship is pending validation.

ItsMii176 committed:
🐛 Fix combobox scroll: always scroll to top on search to show results

ItsMii176 committed:
🚨Fix CSS linting warnings

ItsMii176 committed:
🚸 Replace Dialog with AlertDialog in import summary

Swapped usage of Dialog components for AlertDialog components in ExcelImportSummaryDialog to improve consistency and UI semantics.

ItsMii176 committed:
✨ Add sorting feature to internship queries

Introduces client-side and server-side sorting for internships by organization, location, date, and duration. Updates infinite scroll and query logic to support sorting and ordering parameters, with validation and fallback strategies for relational fields. Adds modular sorting strategies and integrates search-aware sorting for improved results.

ItsMii176 committed:
🚑 Fix import path for getInternshipsWithSearch

ItsMii176 committed:
📝 Add Excel import guide for administrators

ItsMii176 committed:
🚸 Improve error messages for internship approval

Error messages shown when internship approval fails are now more specific, distinguishing between invalid data and duplicate internships.
